ribs/emitters/_gradient_emitter.py

```diff
@@ -0,0 +1,297 @@
+            solutions to evaluate.
+        """
+        if self.archive.empty and self._initial_solutions is not None:
+            return np.empty((0, self.archive.solution_dim))
+
+        if self.archive.empty:
+            parents = np.expand_dims(self.x0, axis=0)
+        else:
+            parents = self.archive.sample_elites(self.batch_size).solution_batch
+
+        if self._use_isolinedd:
+            noise = self._rng.normal(
+                loc=0.0,
+                scale=self.sigma0,
+                size=(self.batch_size, self.solution_dim),
+            ).astype(self.archive.dtype)
+
+            directions = self.archive.sample_elites(
+                self._batch_size).solution_batch - parents
+
+            line_gaussian = self._rng.normal(
+                scale=self._line_sigma,
+                size=(self._batch_size, 1),
+            ).astype(self.archive.dtype)
+
+            sol = parents + line_gaussian * directions
+            sol = np.minimum(np.maximum(sol + noise, self.lower_bounds),
+                             self.upper_bounds)
+        else:
+            noise = self._rng.normal(
+                loc=0.0,
+                scale=self.sigma0,
+                size=(self.batch_size, self.solution_dim),
+            ).astype(self.archive.dtype)
+
+            sol = np.minimum(np.maximum(parents + noise, self.lower_bounds),
+                             self.upper_bounds)
+
+        self._parents = sol
+        return self._parents
+
+    def ask(self):
+        """Samples new solutions from a gradient arborescence parameterized by a
+        multivariate Gaussian distribution.
+
+        The multivariate Gaussian is parameterized by sigma_g.
+
+        This method returns ``batch_size`` solutions, even though one solution
+        is returned via ``ask_dqd``.
+        
+        Returns:
+            (:attr:`batch_size`, :attr:`solution_dim`) array -- a batch of new
+            solutions to evaluate.
+        Raises:
+            RuntimeError: This method was called without first passing gradients
+                with calls to ask_dqd() and tell_dqd().
+        """
+        if self.archive.empty and self._initial_solutions is not None:
+            return self._initial_solutions
+
+        if self._jacobian_batch is None:
+            raise RuntimeError("Please call ask_dqd() and tell_dqd() "
+                               "before calling ask().")
+
+        if self._measure_gradients:
+            noise = self._rng.normal(
+                loc=0.0,
+                scale=self._sigma_g,
+                size=self._jacobian_batch.shape[:2],
+            )
+            noise[:, 0] = np.abs(noise[:, 0])
+            noise = np.expand_dims(noise, axis=2)
+            offsets = np.sum(np.multiply(self._jacobian_batch, noise), axis=1)
+            sols = offsets + self._parents
+        else:
+            # Transform the Jacobian
+            if len(self._jacobian_batch.shape) == 3:
```

Comment:
What causes the shape to be 3D? When is it not 3D?

Reply:
the 3D shape is `(batch, 1 + num_measure, solution_dim)`, where the "1+" is reserved for objective gradient.

For when it's not 3D... I guess I could see a corner case where measure functions are not used for a DQD domain, and the jacob_batch is supposed to be `(batch, 1, solution_dim)` but instead already got squeezed down to `(batch, solution_dim)`. 

AFAIK, 3D shape is the correct behavior, which then gets squeezed down to 2D in the next few lines of ask(). But if it already got reduced to 2D before the if statement check here, maybe there was an aggregation operator that was not supposed to already be called? e.g. something other than obj_grad was summed/averaged along with it.